source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -10,709 +10,1371 @@
+            // Update varsWithSemanticInfo.
+            // 2. If IRStructType:
+            //  2a. Recurse this function with 'decorations that carry semantic info' from parent.
+            // 3. If not IRStructType:
+            //  3a. Emit 'newField' equal to 'oldField', add 'decorations which carry semantic info'.
+            //  3b. Store a mapping from 'oldField' to 'newField' in 'mapFieldToField'. This info is needed to copy between types.
+            for (auto oldField : src->getFields())
+            {
+                // step 1
+                auto oldKey = oldField->getKey();
+                IRSemanticDecoration* fieldSemanticDecoration = parentSemanticDecoration;
+                if (auto oldSemanticDecoration = oldKey->findDecoration<IRSemanticDecoration>())
+                    fieldSemanticDecoration = oldSemanticDecoration;
+
+                IRLayoutDecoration* fieldLayout = parentLayout;
+                if (auto oldLayout = oldKey->findDecoration<IRLayoutDecoration>())
+                {
+                    fieldLayout = oldLayout;
+                }
+                if (fieldSemanticDecoration != parentSemanticDecoration || parentLayout != fieldLayout)
+                    varsWithSemanticInfo.add(oldField);
+
+                // step 2a
+                if (auto structFieldType = as<IRStructType>(oldField->getFieldType()))
+                {
+                    _flattenNestedStructs(builder, dst, structFieldType, fieldSemanticDecoration, fieldLayout, mapFieldToField, varsWithSemanticInfo);
+                    continue;
+                }
+
+                // step 3a
+                auto newKey = builder.createStructKey();
+                copyNameHintAndDebugDecorations(newKey, oldKey);
+
+                auto newField = builder.createStructField(dst, newKey, oldField->getFieldType());
+                copyNameHintAndDebugDecorations(newField, oldField);
+
+                if (fieldSemanticDecoration)
+                    builder.addSemanticDecoration(newKey, fieldSemanticDecoration->getSemanticName(), fieldSemanticDecoration->getSemanticIndex());
+
+                if (fieldLayout)
+                {
+                    IRLayout* oldLayout = fieldLayout->getLayout();
+                    List<IRInst*> instToCopy;
+                    // Only copy certain decorations needed for resolving system semantics
+                    for (UInt i = 0; i < oldLayout->getOperandCount(); i++)
+                    {
+                        auto operand = oldLayout->getOperand(i);
+                        if (as<IRVarOffsetAttr>(operand) || as<IRUserSemanticAttr>(operand) || as<IRSystemValueSemanticAttr>(operand) || as<IRStageAttr>(operand))
+                            instToCopy.add(operand);
+                    }
+                    IRVarLayout* newLayout = builder.getVarLayout(instToCopy);
+                    builder.addLayoutDecoration(newKey, newLayout);
+                }
+                // step 3b
```

Comment:
I think this will insert conflicting fields if a struct has been used in more than 1 field.